MetaMask/metamask-extension
# Keep memstore contents after service worker restarts

Fixes 

#15954 
#16456 

## Explanation

The memstore contents is usually lost when the MetaMask extension window is closed.
The problem however is that when Service Worker restarts, the memstore is also cleared. This should not be the case

This PR intends to solve this problem by copying the controllers in memtore to persistent store and adding a method to manually clear the store of the controllers when the MetaMask extension is closed.

## More Information

* See: #15828 
* 
Blocked By #15558 

## Manual Testing Steps

1. Open [test-dapp](https://metamask.github.io/test-dapp/)
2. Connect to MetaMask
3. Click on any of SIGN buttons
4. MetaMask window should open to the sign page
5. Open developer console and click on Application Tab
6. Stop service worker
7. Service worker should be restarted automatically 
8. The MetaMask sign window should still be open
9. Close the window.
10. Open the window again
11. The window should open to the home screen (not sign screen).

**NOTE**: This may be not fully testable until David's work on #15558 is fully completed. Because as of now, when service worker is restarted the user is logged out.

## Pre-Merge Checklist


### + If there are functional changes:
